app/src/main/java/org/projectbuendia/client/ui/OdkActivityLauncher.java

```diff
@@ -180,6 +188,105 @@ public static void showOdkCollect(
         callingActivity.startActivityForResult(intent, requestCode);
     }
 
+    /**
+     * Loads the xform from the cache and launches ODK using it. Return true if the cache is
+     * available.
+     * @param callingActivity the {@link Activity} requesting the xform; when ODK closes, the user
+     *                        will be returned to this activity
+     * @param uuidToShow      UUID of the form to show
+     * @param requestCode     if >= 0, this code will be returned in onActivityResult() when the
+     *                        activity exits
+     * @param patient         the {@link org.odk.collect.android.model.Patient} that this form entry will
+     *                        correspond to
+     * @param fields          a {@link Preset} object with any form fields that should be
+     *                        pre-populated
+     */
+    private static boolean loadXformFromCache(final Activity callingActivity,
+                                              final String uuidToShow,
+                                              final int requestCode,
+                                              @Nullable final org.odk.collect.android.model.Patient patient,
+                                              @Nullable final Preset fields) {
+        List<OpenMrsXformIndexEntry> entries = getLocalFormEntries();
+        OpenMrsXformIndexEntry formToShow = findUuid(entries, uuidToShow);
+        if (!formToShow.makeFileForForm().exists()) return false;
+
+        LOG.i(String.format("Using form %s from local cache.", uuidToShow));
+        showForm(callingActivity, requestCode, patient, fields, formToShow);
+
+        return true;
+    }
+
+    private static List<OpenMrsXformIndexEntry> getLocalFormEntries() {
+        List<OpenMrsXformIndexEntry> entries = new ArrayList<>();
+
+        final ContentResolver resolver = App.getInstance().getContentResolver();
+        Cursor c = resolver.query(Contracts.Forms.CONTENT_URI, new String[] {Contracts.Forms.UUID,
+            Contracts.Forms.NAME}, null, null, null);
+        try {
+            while (c.moveToNext()) {
+                String uuid = Utils.getString(c, Contracts.Forms.UUID);
+                String name = Utils.getString(c, Contracts.Forms.NAME);
+                long date = 0; // date is not important here
+                entries.add(new OpenMrsXformIndexEntry(uuid, name, date));
+            }
+        } finally {
+            c.close();
+        }
+
+        return entries;
+    }
+
+    private static void handleSyncError(VolleyError error) {
```

Comment:
Suggest renaming this to `handleFetchError`, because we could be fetching a form when we're not syncing.